File: platform/lang-impl/src/com/intellij/codeInsight/documentation/FontSizeMouseWheelListener.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.codeInsight.documentation;

import com.intellij.openapi.editor.ex.EditorSettingsExternalizable;
import com.intellij.openapi.editor.ex.util.EditorUtil;
import com.intellij.openapi.options.FontSize;
import org.jetbrains.annotations.NotNull;

import java.awt.event.MouseWheelEvent;
import java.awt.event.MouseWheelListener;
import java.util.function.Consumer;

/**
 * @deprecated Unused in v2 implementation.
 */
@Deprecated(forRemoval = true)
final class FontSizeMouseWheelListener implements MouseWheelListener {

  private final @NotNull Consumer<? super @NotNull FontSize> mySizeConsumer;

  FontSizeMouseWheelListener(@NotNull Consumer<? super @NotNull FontSize> sizeConsumer) {
    mySizeConsumer = sizeConsumer;
  }

  @Override
  public void mouseWheelMoved(MouseWheelEvent e) {
    if (!EditorSettingsExternalizable.getInstance().isWheelFontChangeEnabled() || !EditorUtil.isChangeFontSize(e)) {
      return;
    }

    int rotation = e.getWheelRotation();
    if (rotation == 0) return;
    int change = Math.abs(rotation);
    boolean increase = rotation <= 0;
    FontSize initial = DocumentationComponent.getQuickDocFontSize();
    FontSize newFontSize = initial;
    for (; change > 0; change--) {
      if (increase) {
        newFontSize = newFontSize.larger();
      }
      else {
        newFontSize = newFontSize.smaller();
      }
    }
    if (newFontSize == initial) {
      return;
    }

    DocumentationComponent.setQuickDocFontSize(newFontSize);
    DocFontSizePopup.update(newFontSize);
    mySizeConsumer.accept(newFontSize);
  }
}

File: platform/lang-impl/src/com/intellij/lang/documentation/ide/ui/FontSizeMouseWheelListener.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.lang.documentation.ide.ui

import com.intellij.openapi.editor.ex.EditorSettingsExternalizable
import com.intellij.openapi.editor.ex.util.EditorUtil
import com.intellij.openapi.options.FontSize
import com.intellij.ui.FontSizeModel
import java.awt.event.MouseWheelEvent
import java.awt.event.MouseWheelListener
import kotlin.math.abs

internal class FontSizeMouseWheelListener(
  private val model: FontSizeModel<FontSize>,
) : MouseWheelListener {

  override fun mouseWheelMoved(e: MouseWheelEvent) {
    if (!EditorSettingsExternalizable.getInstance().isWheelFontChangeEnabled || !EditorUtil.isChangeFontSize(e)) {
      // TODO this check must be done outside this listener: listener should not even be registered if this case
      return
    }
    val rotation = e.wheelRotation
    if (rotation == 0) {
      return
    }
    var newFontSize = model.value
    val increase = rotation <= 0
    repeat(abs(rotation)) {
      newFontSize = if (increase) {
        newFontSize.larger()
      }
      else {
        newFontSize.smaller()
      }
    }
    model.value = newFontSize
  }
}
